synapse/lib/editor.py

```diff
@@ -0,0 +1,998 @@
+            for addname, addvalu, addinfo in propadds:
+                await self.editor.addNode(addname, addvalu, norminfo=addinfo)
+
+        return True
+
+    async def pop(self, name):
+
+        prop = self.form.prop(name)
+        if prop is None:
+            mesg = f'No property named {name}.'
+            await self.editor._raiseOnStrict(s_exc.NoSuchProp, mesg, name=name, form=self.form.name)
+            return False
+
+        (valu, layr) = self.getWithLayer(name, defv=s_common.novalu)
+        if valu is s_common.novalu:
+            return False
+
+        if prop.info.get('ro'):
+            mesg = f'Property is read only: {prop.full}.'
+            await self.editor._raiseOnStrict(s_exc.ReadOnlyProp, mesg, name=prop.full)
+            return False
+
+        self.props.pop(name, None)
+
+        if layr == 0:
+            self.propdels.add(name)
+
+        if self.multilayer:
+            if layr > 0 or (self.node is not None and self.node.hasInLayers(name, strt=1)):
+                self.proptombs.add(name)
+
+        return True
+
+    async def getSetOps(self, name, valu, norminfo=None):
+        prop = self.form.props.get(name)
+        if prop is None:
+            return ()
+
+        retn = await self._set(prop, valu, norminfo=norminfo)
+        if retn is False:
+            return ()
+
+        (valu, norminfo) = retn
+        ops = []
+
+        propform = self.editor.view.core.model.form(prop.type.name)
+        if propform is not None:
+            ops.append(self.editor.getAddNodeOps(propform.name, valu, norminfo=norminfo))
+
+        # TODO can we mandate any subs are returned pre-normalized?
+        propsubs = norminfo.get('subs')
+        if propsubs is not None:
+            for subname, subvalu in propsubs.items():
+                full = f'{prop.name}:{subname}'
+                if self.form.props.get(full) is not None:
+                    ops.append(self.getSetOps(full, subvalu))
+
+        propadds = norminfo.get('adds')
+        if propadds is not None:
+            for addname, addvalu, addinfo in propadds:
+                ops.append(self.editor.getAddNodeOps(addname, addvalu, norminfo=addinfo))
+
+        return ops
+
+    async def delete(self):
+        if self.node is None or self.istomb():
+            return
+
+        if self.node.sodes[0].get('valu') is not None:
+            self.delnode = True
+
+        for sode in self.node.sodes[1:]:
+            if sode.get('valu') is not None:
+                self.tombnode = True
+                return
+            elif sode.get('antivalu') is not None:
+                return
+
+class NodeEditor:
+    '''
+    A NodeEditor allows tracking node edits with subs/deps as a transaction.
+    '''
+    def __init__(self, view, user, runt=None, strict=True):
+        self.runt = runt
+        self.user = user
+        self.view = view
+        self.protonodes = {}
+        self.maxnodes = view.core.maxnodes
+        self.strict = strict
```

Comment:
Do we still need a `strict` distinction? That was originally added for the feeds concept from 0.2.x to facilitate ingests not-failing on bad data. I think the only ingest we support now is `syn.nodes` though.

Reply:
The `syn.nodes` ingest still uses strict=False, I wouldn't be opposed to removing strict=False though

Reply:
I suppose addNodes could just catch errors itself

Reply:
The storm merge command also uses strict=False, but I'm not sure that it should be doing that?

Reply:
Removed strict from the editor and adjusted view.addNodes to catch and log exceptions (and warn to the runtime if there is one)